pkg/loadbalancer/reconciler/termination.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/cilium/cilium/pkg/netns"
+	"github.com/cilium/cilium/pkg/rate"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// SocketTerminationCell runs a background job that monitors the backends table for
+// unhealthy and deleted backends and terminates UDP sockets connected to these backends
+// to signal to the application that the destination has become unreachable.
+var SocketTerminationCell = cell.Module(
+	"socket-termination",
+	"Terminates sockets connected to deleted backends",
+
+	// Provide the default implementations for accessing network namespaces and
+	// destroying sockets. This allows the tests to use "registerSocketTermination"
+	// directly and provide custom implementations of these.
+	cell.ProvidePrivate(
+		func() netnsOps {
+			return netnsOps{
+				current: netns.Current,
+				do:      (*netns.NetNS).Do,
+				all:     netns.All,
+			}
+		},
+		func(log *slog.Logger) sockets.SocketDestroyer {
+			return socketDestroyer{log}
+		},
+	),
+
+	cell.Invoke(registerSocketTermination),
+)
+
+type socketTerminationParams struct {
+	cell.In
+
+	JobGroup  job.Group
+	DB        *statedb.DB
+	Backends  statedb.Table[*lb.Backend]
+	Log       *slog.Logger
+	Config    lb.Config
+	ExtConfig lb.ExternalConfig
+	LBMaps    maps.LBMaps
+
+	SocketDestroyer sockets.SocketDestroyer
+	NetNSOps        netnsOps
+
+	// TestSyncChan is used by the tests to synchronize with the job so it
+	// knows when to delete the backend.
+	TestSyncChan testSyncChan   `optional:"true"`
+	TestConfig   *lb.TestConfig `optional:"true"`
+}
+
+type testSyncChan chan struct{}
+
+// netnsOps captures the operations performed with network namespaces.
+// This allows the tests to inject their own implementation.
+type netnsOps struct {
+	current func() (*netns.NetNS, error)
+	do      func(*netns.NetNS, func() error) error
+	all     func() (iter.Seq2[string, *netns.NetNS], <-chan error)
+}
+
+type socketDestroyer struct {
+	log *slog.Logger
+}
+
+func (sd socketDestroyer) Destroy(filter sockets.SocketFilter) error {
+	return sockets.Destroy(sd.log, filter)
+}
+
+func registerSocketTermination(p socketTerminationParams) error {
+	if !p.Config.EnableExperimentalLB {
+		return nil
+	}
+
+	if p.SocketDestroyer == nil {
+		// To make the load-balancer cell easier to use in tests we don't require that
+		// SocketDestroyer is always provided.
+		if p.TestConfig == nil {
+			return fmt.Errorf("SocketDestroyer not provided and not running in tests")
+		}
+		return nil
+	}
+
+	if !(p.ExtConfig.EnableSocketLB || p.ExtConfig.BPFSocketLBHostnsOnly) {
+		return nil
+	}
+
+	p.JobGroup.Add(
+		job.OneShot(
+			"socket-termination",
+			func(ctx context.Context, h cell.Health) error {
+				return socketTerminationLoop(p, ctx, h)
+			},
+		))
+
+	return nil
+}
+
+func socketTerminationLoop(p socketTerminationParams, ctx context.Context, health cell.Health) error {
+	wtxn := p.DB.WriteTxn(p.Backends)
+	changeIter, err := p.Backends.Changes(wtxn)
+	wtxn.Commit()
```

Comment:
why do we commit if it fails?

Reply:
simpler code, nothing happens if it fails, shouldn't really never fail.